pysmiles/smiles_helper.py

```diff
@@ -342,37 +347,70 @@ def bonds_missing(mol, node_idx, use_order=True):
     """
     bonds = _bonds(mol, node_idx, use_order)
     bonds += mol.nodes[node_idx].get('hcount', 0)
-    valence = _valence(mol, node_idx, bonds)
-    return int(valence - bonds)
+
+    val = valence(mol.nodes[node_idx])
+    val = [v for v in val if v >= bonds] or [0]
+    return int(max(val[0] - bonds, 0))
 
 
-def _valence(mol, node_idx, minimum=0):
+def valence(atom):
     """
-    Returns the valence of the specified node. Since some elements can have
-    multiple valences, give the smallest one that is more than `minimum`.
+    Returns the valence of the atom. Since some elements can have
+    multiple valences, The valence is returned as list.
 
     Parameters
     ----------
-    mol : nx.Graph
-        The molecule.
-    node_idx : hashable
-        The node to look at. Should be in mol.
-    minimum : int
-        The minimum value of valence.
+    atom: dict
 
     Returns
     -------
-    int
-        The smallest valence of node more than `minimum`.
+    list[int]
+        The valences for the given atom.
     """
-    element = mol.nodes[node_idx].get('element', '').capitalize()
-    if element not in VALENCES:
-        return 0
-    val = VALENCES.get(element)
-    try:
-        val = min(filter(lambda a: a >= minimum, val))
-    except ValueError:  # More bonds than possible
-        val = max(val)
+    element = atom.get("element", '*')
+    if element == '*':
+        electrons = 0
+    else:
+        electrons = PTE[element.capitalize()]['AtomicNumber']
+    electrons -= atom.get('charge', 0)
+
+    if electrons < 0:
+        raise ValueError(f"Atom {atom} has a negative number of electrons: {electrons}")
+
+    # Let's start by filling complete shells:
+    for idx, shell in enumerate(ORBITAL_SIZES):
+        shell_size = sum(shell)
+        if shell_size <= electrons:
+            electrons -= shell_size
+        else:
+            break
+    else:  # nobreak
+        raise ValueError(f'Too many electrons for sanity for {atom}: {electrons+sum(map(sum, ORBITAL_SIZES))}')
+
+    # Any electrons we have leftover we distribute over the orbitals. First 1
+    # electron in each, then we start making pairs. The resulting valence will
+    # be the number of unpaired electrons. Added bonus/complication: electrons
+    # in pairs we can excite to higher shells, increasing the number of unpaired
+    # electrons
+    shell = ORBITAL_SIZES[idx]
+    shell_size = sum(shell)
+    single_electrons, paired_electrons = divmod(electrons, shell_size//2)
+    if single_electrons == 0:
+        single_electrons, paired_electrons, empty_orbitals = paired_electrons, 0, shell_size//2 - paired_electrons
```

Comment:
what is the purpose of empty_orbitals here? why not say `single_electrons = shell_size//2` and `paired_electrons = 0`

Reply:
It's mostly for later (YAGNI!), if we want to be able to adjust the charge to fix the valence of atoms.